src/common/src/types/to_binary.rs

```diff
@@ -78,8 +79,45 @@ implement_using_to_sql! {
     { Timestamptz, Timestamptz, |x: &Timestamptz| x.to_datetime_utc() }
 }
 
+impl ToBinary for ListRef<'_> {
+    fn to_binary_with_type(&self, ty: &DataType) -> Result<Bytes> {
+        // Reference: Postgres code `src/backend/utils/adt/arrayfuncs.c`
+        // https://github.com/postgres/postgres/blob/c1c09007e219ae68d1f8428a54baf68ccc1f8683/src/backend/utils/adt/arrayfuncs.c#L1548
+        use crate::row::Row;
+        let element_ty = match ty {
+            DataType::List(ty) => ty.as_ref(),
+            _ => unreachable!(),
+        };
+        if matches!(element_ty, DataType::List(_)) {
+            bail_not_implemented!(
+                issue = 7949,
+                "list with 2 or more dimensions is not supported"
+            )
+        }
+        let mut buf = BytesMut::new();
+        buf.put_i32(1); // Number of dimensions (must be 1)
+        buf.put_i32(1); // Has nulls?
+        buf.put_i32(element_ty.to_oid()); // Element type
+        buf.put_i32(self.len() as i32); // Length of 1st dimension
+        buf.put_i32(0); // Offset of 1st dimension
```

Comment:
SQL array starts at 1 rather than 0.
```
test=# select array_send(array[2, 3, null]);
                                     array_send                                     
------------------------------------------------------------------------------------
 \x000000010000000100000017000000030000000100000004000000020000000400000003ffffffff
(1 row)
```